### PR TITLE
LPS-188416 Add Pagination support to HeadlessBuilder endpoints

### PR DESCRIPTION
**Pull request by LuismiBarcos**
Related ticket [LPS-188416](https://liferay.atlassian.net/browse/LPS-188416)
____
## Dependencies
This PR should be sent once [this PR](https://github.com/liferay-headless/liferay-portal/pull/1336) would be merged

## Purpose of the PR
This PR adds support for Pagination for the Headless Builder REST Application endpoints. 
With this, a user could now do requests with paginated params like in other Liferay collections endpoints.

## How to test it
It is necessary to deploy the following modules:
* headless-builder-api
* headless-builder-impl
* portal-vulcan-impl

### CODE REVIEW

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:relevant

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/writer/interceptor/PageEntityExtensionWriterInterceptor.java [hunk 1]
```diff
@@ -50,7 +50,10 @@ public class PageEntityExtensionWriterInterceptor implements WriterInterceptor {
 	public void aroundWriteTo(WriterInterceptorContext writerInterceptorContext)
 		throws IOException {
 
-		if (Page.class.isAssignableFrom(writerInterceptorContext.getType())) {
+		if (Page.class.isAssignableFrom(writerInterceptorContext.getType()) &&
```
Comment (4lejandrito) [inline modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/writer/interceptor/PageEntityExtensionWriterInterceptor.java:53]: Can this break a preexisting use case? The condition is now more restrictive.

**Review by 4lejandrito — COMMENTED** [review]
Just a small comment for now.

File: modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/writer/interceptor/PageEntityExtensionWriterInterceptor.java (hunk 1 above)
Comment (LuismiBarcos) [reply modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/jaxrs/writer/interceptor/PageEntityExtensionWriterInterceptor.java:53]: Well, the condition I am adding is just to ensure that the next line could be executed and will not throw an NPE. I am not making the condition more restrictive, just avoiding NPE.

If something doesn't work now, it should be "exploding" actually 😅 
With this, will just continue with the execution

File: modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/application/resource/test/HeadlessBuilderResourceTest.java [hunk 2]
```diff
@@ -150,22 +143,18 @@ public void testEndpointReturnsProperSchema() throws Exception {
...
-
-		String actual = StringUtil.removeSubstrings(
-			StringUtil.read(httpURLConnection.getInputStream()),
-			StringPool.NEW_LINE, StringPool.SPACE);
-
-		String expected = String.format(
-			"[{\"name\":\"%s\"}]", _OBJECT_FIELD_VALUE);
-
-		Assert.assertEquals(expected, actual);
+		JSONAssert.assertEquals(
```
Comment (ccorreagg) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/application/resource/test/HeadlessBuilderResourceTest.java:184]: Maybe better to validate the whole page instead (with items, page number, etc)

File: modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/application/resource/test/HeadlessBuilderResourceTest.java [hunk 3]
```diff
@@ -136,9 +136,43 @@ public void test() throws Exception {
...
+
+		JSONObject jsonObject = HTTPTestUtil.invoke(
+			null,
+			String.format(
+				"%s?page=%d&pageSize=%d",
+				_apiApplication1.getBaseURL() + _API_APPLICATION_PATH_1, page,
+				pageSize),
+			Http.Method.GET);
+
+		Assert.assertEquals(lastPage, jsonObject.get("lastPage"));
```
Comment (ccorreagg) [inline modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/application/resource/test/HeadlessBuilderResourceTest.java:166]: I think we could use `JSONAssert.assertEquals` instead as it contains the `JSONCompareMode` argument to modify the way we want to compare (for example, without having into account the order)

File: modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/application/resource/test/HeadlessBuilderResourceTest.java (hunk 2 above)
Comment (ccorreagg) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/application/resource/test/HeadlessBuilderResourceTest.java:184]: Edit: I didn't realise you were testing in another test, but maybe it is also good to include it. Something like:

```
		JSONAssert.assertEquals(
			JSONUtil.put(
				"items", JSONUtil.put(
					JSONUtil.put("name", _OBJECT_FIELD_VALUE)
				)
			).put(
				"lastPage", 1
			).put(
				"page", 1
			).put(
				"pageSize", 20
			).put(
				"totalCount", 1
			).toString(),
			jsonObject.toString(),
			JSONCompareMode.LENIENT
```

Comment (LuismiBarcos) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/application/resource/test/HeadlessBuilderResourceTest.java:192]: Done

File: modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/application/resource/test/HeadlessBuilderResourceTest.java (hunk 3 above)
Comment (LuismiBarcos) [reply modules/apps/headless/headless-builder/headless-builder-test/src/testIntegration/java/com/liferay/headless/builder/application/resource/test/HeadlessBuilderResourceTest.java]: Done

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Review by ccorreagg — APPROVED** [review]
LGTM!

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by ccorreagg** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:relevant

**Comment by 4lejandrito** [comment]
Rebased on top of https://github.com/brianchandotcom/liferay-portal/pull/137697.

**Comment by 4lejandrito** [comment]
ci:test:sf

**Comment by 4lejandrito** [comment]
ci:test:relevant

**Comment by 4lejandrito** [comment]
Rebased from https://github.com/brianchandotcom/liferay-portal/pull/137702.

**Comment by 4lejandrito** [comment]
ci:test:relevant

**Comment by 4lejandrito** [comment]
ci:test:sf

**Comment by 4lejandrito** [comment]
Rebased from Brian's.

**Comment by 4lejandrito** [comment]
See https://github.com/brianchandotcom/liferay-portal/pull/137754.

**Pull request closed without merging**